[SPARK-26379][SS][BRANCH-2.3] Use dummy TimeZoneId to avoid UnresolvedException in CurrentBatchTimestamp

HeartSaVioR opened the pull request:
## What changes were proposed in this pull request?

Spark replaces `CurrentTimestamp` with `CurrentBatchTimestamp`.
However, `CurrentBatchTimestamp` is `TimeZoneAwareExpression` while `CurrentTimestamp` isn't.
Without TimeZoneId, `CurrentBatchTimestamp` becomes unresolved and raises `UnresolvedException`.

Since `CurrentDate` is `TimeZoneAwareExpression`, there is no problem with `CurrentDate`.

## How was this patch tested?

Pass the Jenkins with the updated test cases.

HeartSaVioR committed:
[SPARK-26379][SS] Fix issue on adding current_timestamp/current_date to streaming query

## What changes were proposed in this pull request?

This patch proposes to fix issue on adding `current_timestamp` / `current_date` with streaming query.

The root reason is that Spark transforms `CurrentTimestamp`/`CurrentDate` to `CurrentBatchTimestamp` in MicroBatchExecution which makes transformed attributes not-yet-resolved. They will be resolved by IncrementalExecution.
(In ContinuousExecution, Spark doesn't allow using `current_timestamp` and `current_date` so it has been OK.)

It's OK for DataSource V1 sink because it simply leverages transformed logical plan and don't evaluate until they're resolved, but for DataSource V2 sink, Spark tries to extract the schema of transformed logical plan in prior to IncrementalExecution, and unresolved attributes will raise errors.

This patch fixes the issue via having separate pre-resolved logical plan to pass the schema to StreamingWriteSupport safely.

## How was this patch tested?

Added UT.

Closes #23609 from HeartSaVioR/SPARK-26379.

Authored-by: Jungtaek Lim (HeartSaVioR) <kabhwan@gmail.com>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented:
Thank you, @HeartSaVioR . :)

SparkQA commented:
**[Test build #101696 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101696/testReport)** for PR 23656 at commit [`cec965c`](https://github.com/apache/spark/commit/cec965c87f4f4289dd2b61fa00832eaaef5c5cda).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101696/
Test FAILed.

dongjoon-hyun commented:
We will revisit this PR because the original PR needs a more verification and a followup PR is expected.
- https://github.com/apache/spark/pull/23609

HeartSaVioR commented:
FYI: follow-up PR is available in #23660 submitted by Dongjoon.

dongjoon-hyun committed:
[SPARK-26379][SS][FOLLOWUP] Use dummy TimeZoneId to avoid UnresolvedException in CurrentBatchTimestamp

## What changes were proposed in this pull request?

Spark replaces `CurrentTimestamp` with `CurrentBatchTimestamp`.
However, `CurrentBatchTimestamp` is `TimeZoneAwareExpression` while `CurrentTimestamp` isn't.
Without TimeZoneId, `CurrentBatchTimestamp` becomes unresolved and raises `UnresolvedException`.

Since `CurrentDate` is `TimeZoneAwareExpression`, there is no problem with `CurrentDate`.

This PR reverts the [previous patch](https://github.com/apache/spark/pull/23609) on `MicroBatchExecution` and fixes the root cause.

## How was this patch tested?

Pass the Jenkins with the updated test cases.

Closes #23660 from dongjoon-hyun/SPARK-26379.

Authored-by: Dongjoon Hyun <dongjoon@apache.org>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>

dongjoon-hyun commented:
Hi, @HeartSaVioR . Could you update this PR with the final content?

HeartSaVioR commented:
@dongjoon-hyun Yes, I've just cherry-picked the FOLLOW-UP commit into PR branch.

SparkQA commented:
**[Test build #101734 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101734/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).

dongjoon-hyun commented:
Thank you, @HeartSaVioR . It's enough since this will be merged as a single commit.

dongjoon-hyun commented:
@HeartSaVioR . Could you update the PR title and description, too?

SparkQA commented:
**[Test build #101734 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101734/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101734/
Test FAILed.

HeartSaVioR commented:
@dongjoon-hyun Sorry for the late. Just updated the title and content of PR.

HeartSaVioR commented:
Test failure looks not relevant. Retest this, please.

maropu commented:
yea, that's known flaky tests in branch-2.3..

SparkQA commented:
**[Test build #101740 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101740/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).

SparkQA commented:
**[Test build #101740 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101740/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101740/
Test FAILed.

maropu commented:
retest this please

SparkQA commented:
**[Test build #101743 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101743/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).

dongjoon-hyun commented:
Thank you, @HeartSaVioR and @maropu .

dongjoon-hyun reviewed (approved):
+1, LGTM (pending Jenkins).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7474/
Test PASSed.

dongjoon-hyun commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7478/
Test PASSed.

SparkQA commented:
**[Test build #101747 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101747/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).

SparkQA commented:
**[Test build #101747 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101747/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

HeartSaVioR commented:
retest this, please

SparkQA commented:
**[Test build #101749 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101749/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).

SparkQA commented:
**[Test build #101749 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101749/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101749/
Test FAILed.

HeartSaVioR commented:
Retest this please

SparkQA commented:
**[Test build #101758 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101758/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).

maropu commented:
super flaky....

HeartSaVioR commented:
Yeah... If the branch should be maintained further, it may need to fix the flaky tests... As it seems to be known flakiness, are we aware about the fix in upper version lines?

SparkQA commented:
**[Test build #101758 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101758/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101758/
Test FAILed.

dongjoon-hyun commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7487/
Test PASSed.

SparkQA commented:
**[Test build #101763 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101763/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #101764 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101764/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7488/
Test PASSed.

dongjoon-hyun commented:
Retest this please.

dongjoon-hyun commented:
@HeartSaVioR and @maropu . I triggered Jenkins three times intentionally to pass the Jenkins.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7489/
Test PASSed.

SparkQA commented:
**[Test build #101765 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101765/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101765/
Test FAILed.

SparkQA commented:
**[Test build #101764 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101764/testReport)** for PR 23656 at commit [`52ac627`](https://github.com/apache/spark/commit/52ac627589db4ad63bf959406ad70b9d8790f40a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101764/
Test FAILed.

HeartSaVioR commented:
2 failures, 1 passes. It looks like kind of serious one if we will maintain branch-2.3 forward.

dongjoon-hyun commented:
@HeartSaVioR . Yes. It's true and `branch-2.3` will be EOF in several months.

dongjoon-hyun commented:
Merged to branch-2.3.

maropu commented:
@HeartSaVioR I'm not familiar with the SS part though, your other pr (https://github.com/apache/spark/pull/23634) is not related to the failures? It seems the failures are of the outer joins tests...

HeartSaVioR commented:
If #23634 was the fix for tests they would be flaky for branch-2.4 / master as well. They don't seem to be flaky in master branch.

maropu commented:
Aha, I see.
